src/main/java/org/jitsi/videobridge/cc/AdaptiveTrackProjection.java

```diff
@@ -0,0 +1,307 @@
+     *
+     * Note that we keep a {@link WeakReference} instead of a reference to allow
+     * the channel/stream/etc objects to be de-allocated in case the sending
+     * participant leaves the conference.
+     */
+    private final WeakReference<MediaStreamTrackDesc> weakSource;
+
+    private final long targetSsrc;
+
+    /**
+     * The payload specific track projection context that's responsible for
+     * rewriting the packets of a projected track.
+     */
+    private AdaptiveTrackProjectionContext context;
+
+    /**
+     * The payload type that was used to determine the {@link #context} type.
+     */
+    private int contextPayloadType = -1;
+
+    /**
+     * The ideal quality index for this track projection.
+     */
+    private int idealIndex = RTPEncodingDesc.SUSPENDED_INDEX;
+
+    /**
+     * The target quality index for this track projection.
+     */
+    private int targetIndex = RTPEncodingDesc.SUSPENDED_INDEX;
+
+    /**
+     * Ctor.
+     *
+     * @param source the {@link MediaStreamTrackDesc} that owns the packets
+     * that this instance filters.
+     */
+    AdaptiveTrackProjection(@NotNull MediaStreamTrackDesc source)
+    {
+        weakSource = new WeakReference<>(source);
+        targetSsrc = source.getRTPEncodings()[0].getPrimarySSRC();
+    }
+
+    /**
+     * @return the {@link MediaStreamTrackDesc} that owns the packets that this
+     * instance filters. Note that this may return null.
+     */
+    public MediaStreamTrackDesc getSource()
+    {
+        return weakSource.get();
+    }
+
+    /**
+     * @return the ideal quality for this track projection.
+     */
+    int getIdealIndex()
+    {
+        return idealIndex;
+    }
+
+    /**
+     * Update the ideal quality for this track projection.
+     *
+     * @param value the ideal quality for this track projection.
+     */
+    void setIdealIndex(int value)
+    {
+        idealIndex = value;
+    }
+
+    /**
+     * Gets the target index value for this track projection.
+     *
+     * @return the target index value for this track projection.
+     */
+    int getTargetIndex()
+    {
+        return targetIndex;
+    }
+
+    /**
+     * Sets the target index value for this track projection.
+     *
+     * @param value the new target index value for this track projection.
+     */
+    void setTargetIndex(int value)
+    {
+        targetIndex = value;
+    }
+
+    /**
+     * Determines whether an RTP packet needs to be accepted or not.
+     *
+     * @param rtpPacket the RTP packet to determine whether to accept or not.
+     * @return true if the packet is accepted, false otherwise.
+     */
+    public boolean accept(@NotNull RawPacket rtpPacket)
+    {
+        AdaptiveTrackProjectionContext
+            contextCopy = getContext(RawPacket.getPayloadType(rtpPacket));
+
+        int targetIndexCopy = targetIndex;
+        boolean accept = contextCopy.accept(rtpPacket, targetIndexCopy);
```

Comment:
could we check if `targetIndexCopy == SUSPENDED_INDEX` and, if so, return false immediately?

Reply:
We want to let the context know that the stream has been suspended so that it can raise the needsKeyframe flag and also allow it to compute a sequence number delta when the target becomes > -1.

Reply:
i see, so this is how a context 'learns' it has been suspended?  would it be easier to notify it only when targetIndex changes, rather than for every packet?

Reply:
Sure, so we could, for example, introduce an `AdaptiveTrackProjectionContext.targetIndex` field in the context classes. This would effectively make the `AdaptiveTrackProjection.targetIndex` field redundant and it would have to be removed.

The only annoyance then would be that the track projection owns the `idealIndex` and the projection context owns the `targetIndex`. I can implement this change, but I also kind of like how we're storing the two indices in the same place now.

Unless you have something else in mind?

Reply:
Ok, I see what you mean.  I think it's fine as it is.